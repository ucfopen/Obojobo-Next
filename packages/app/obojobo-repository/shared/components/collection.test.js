jest.mock('short-uuid')
jest.mock('./collection-menu-hoc', () => props => {
	return <mock-CollectionMenu {...props}>{props.children}</mock-CollectionMenu>
})

import React from 'react'
import Collection from './collection'
import CollectionMenu from './collection-menu-hoc'

import { create, act } from 'react-test-renderer'

describe('Collection', () => {
	const mockShortFromUUID = jest.fn()

	let defaultProps
	let short

	beforeEach(() => {
		jest.resetAllMocks()
		jest.useFakeTimers()

		defaultProps = {
			id: 'mockCollectionId',
			title: 'Mock Collection Title',
			hasMenu: true
		}

		short = require('short-uuid')
		mockShortFromUUID.mockReturnValue('mockCollectionShortId')
		short.mockReturnValue({
			fromUUID: mockShortFromUUID
		})
	})

	test('renders with expected standard props', () => {
		const component = create(<Collection {...defaultProps} />)

		expect(mockShortFromUUID).toHaveBeenCalledTimes(1)
		expect(mockShortFromUUID).toHaveBeenCalledWith('mockCollectionId')

		expect(component.root.findAllByType(CollectionMenu).length).toBe(0)

		expect(component.root.children[0].props.className).toBe(
			'repository--collection-icon is-not-open'
		)

		const mainChildComponent = component.root.children[0].children[0]
		expect(mainChildComponent.type).toBe('button')

		expect(component.toJSON()).toMatchSnapshot()
	})

	test('renders with expected standard props but hasMenu=false', () => {
		defaultProps.hasMenu = false
		const component = create(<Collection {...defaultProps} />)

		expect(mockShortFromUUID).toHaveBeenCalledTimes(1)
		expect(mockShortFromUUID).toHaveBeenCalledWith('mockCollectionId')

		expect(component.root.findAllByType(CollectionMenu).length).toBe(0)

		expect(component.root.children[0].props.className).toBe(
			'repository--collection-icon is-not-open'
		)

		const mainChildComponent = component.root.children[0].children[0]
		expect(mainChildComponent.type).toBe('a')
		expect(mainChildComponent.props.href).toBe(
			'/collections/mock-collection-title-mockCollectionShortId'
		)

		expect(component.toJSON()).toMatchSnapshot()
	})

	test('clicking the main child component when hasMenu=true renders the collection menu', () => {
		const reusableComponent = <Collection {...defaultProps} />
		let component
		act(() => {
			component = create(reusableComponent)
		})

		expect(component.root.findAllByType(CollectionMenu).length).toBe(0)

		const mockClickEvent = {
			preventDefault: jest.fn()
		}
		act(() => {
			component.root.children[0].children[0].props.onClick(mockClickEvent)
			component.update(reusableComponent)
		})
		expect(mockClickEvent.preventDefault).toHaveBeenCalledTimes(1)
		expect(component.root.findAllByType(CollectionMenu).length).toBe(1)
	})

	test('the collection menu is not rendered after onMouseLeave is called', () => {
		const reusableComponent = <Collection {...defaultProps} />
		let component
		act(() => {
			component = create(reusableComponent)
		})

		expect(component.root.findAllByType(CollectionMenu).length).toBe(0)

		act(() => {
			const mockClickEvent = {
				preventDefault: jest.fn()
			}
			component.root.children[0].children[0].props.onClick(mockClickEvent)
			component.update(reusableComponent)
		})
		expect(component.root.findAllByType(CollectionMenu).length).toBe(1)

		act(() => {
			component.root.children[0].props.onMouseLeave()
			component.update(reusableComponent)
		})

		expect(component.root.findAllByType(CollectionMenu).length).toBe(0)
	})

	test('keyboard focus and blur close the collection menu as necessary', () => {
		const reusableComponent = <Collection {...defaultProps} />
		let component
		act(() => {
			component = create(reusableComponent)
		})

		expect(component.root.findAllByType(CollectionMenu).length).toBe(0)

		act(() => {
			component.root.children[0].props.onFocus()
		})

		expect(window.clearTimeout).toHaveBeenCalledTimes(1)

		// bring up the collection menu now so we can make sure it's gone later
		act(() => {
			const mockClickEvent = {
				preventDefault: jest.fn()
			}
			component.root.children[0].children[0].props.onClick(mockClickEvent)
			component.update(reusableComponent)
		})
		expect(component.root.findAllByType(CollectionMenu).length).toBe(1)

		act(() => {
			component.root.children[0].props.onBlur()

			expect(window.setTimeout).toHaveBeenCalledTimes(1)
			jest.runAllTimers()

			component.update(reusableComponent)
		})

		expect(component.root.findAllByType(CollectionMenu).length).toBe(0)
	})
})
