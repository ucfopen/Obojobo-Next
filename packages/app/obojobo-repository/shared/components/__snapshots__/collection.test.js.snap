// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection renders with expected standard props 1`] = `
<div
  className="repository--collection-icon is-not-open"
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseLeave={[Function]}
>
  <button
    onClick={[Function]}
  >
    <div
      className="repository--collection-icon--image"
    >
      <svg
        height="80"
        preserveAspectRatio="xMidYMid meet"
        version="1.1"
        viewBox="0 0 80 80"
        width="80"
        xmlns="http://www.w3.org/2000/svg"
      >
        <defs>
          <pattern
            dangerouslySetInnerHTML={
              Object {
                "__html": "<image xlink:href=\\"/library/module-icon/mockCollectionId\\" width=\\"150\\" height=\\"150\\" />",
              }
            }
            height="150"
            id="collection-img-id-mockCollectionId"
            patternUnits="userSpaceOnUse"
            width="150"
          />
        </defs>
        <rect
          fill="#e8e3eb"
          height="80"
          rx="5"
          width="80"
        />
        <path
          className="repository--collection-icon--pattern-path"
          d="m20.092289,37.657358q2.2001,-1.270228 4.400197,0l12.100543,6.986251q2.200097,1.270228 2.200097,3.810684l0,13.972502q0,2.540453 -2.200097,3.810681l-12.100543,6.986251q-2.200097,1.270228 -4.400197,0l-12.100543,-6.986251q-2.200097,-1.270228 -2.200097,-3.810681l0,-13.972502q0,-2.540456 2.200097,-3.810684l12.100543,-6.986251z"
          fill="url(#collection-img-id-mockCollectionId)"
        />
        <path
          d="m55.507514,37.65736q2.200094,-1.270229 4.400197,0l12.100545,6.986249q2.200094,1.270229 2.200094,3.81068l0,13.972507q0,2.54045 -2.200094,3.81068l-12.100545,6.986249q-2.200103,1.270229 -4.400197,0l-12.100545,-6.986249q-2.200094,-1.270229 -2.200094,-3.81068l0,-13.972507q0,-2.54045 2.200094,-3.81068l12.100545,-6.986249z"
          fill="#9482a6"
        />
        <path
          d="m37.861174,6.776266q2.200094,-1.270229 4.400197,0l12.100545,6.986249q2.200094,1.270229 2.200094,3.81068l0,13.972507q0,2.54045 -2.200094,3.81068l-12.100545,6.986249q-2.200103,1.270229 -4.400197,0l-12.100545,-6.986249q-2.200094,-1.270229 -2.200094,-3.81068l0,-13.972507q0,-2.54045 2.200094,-3.81068l12.100545,-6.986249z"
          fill="#b09ec3"
        />
      </svg>
    </div>
    <div
      className="repository--collection-icon--menu-control-button menu-control-button"
    >
      <svg
        height="32px"
        viewBox="0 0 32 32"
        width="32px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M4,10h24c1.104,0,2-0.896,2-2s-0.896-2-2-2H4C2.896,6,2,6.896,2,8S2.896,10,4,10z M28,14H4c-1.104,0-2,0.896-2,2  s0.896,2,2,2h24c1.104,0,2-0.896,2-2S29.104,14,28,14z M28,22H4c-1.104,0-2,0.896-2,2s0.896,2,2,2h24c1.104,0,2-0.896,2-2  S29.104,22,28,22z"
        />
      </svg>
    </div>
    <div
      className="repository--collection-icon--title"
    >
      Mock Collection Title
    </div>
  </button>
</div>
`;

exports[`Collection renders with expected standard props but hasMenu=false 1`] = `
<div
  className="repository--collection-icon is-not-open"
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseLeave={[Function]}
>
  <a
    href="/collections/mock-collection-title-mockCollectionShortId"
  >
    <div
      className="repository--collection-icon--image"
    >
      <svg
        height="80"
        preserveAspectRatio="xMidYMid meet"
        version="1.1"
        viewBox="0 0 80 80"
        width="80"
        xmlns="http://www.w3.org/2000/svg"
      >
        <defs>
          <pattern
            dangerouslySetInnerHTML={
              Object {
                "__html": "<image xlink:href=\\"/library/module-icon/mockCollectionId\\" width=\\"150\\" height=\\"150\\" />",
              }
            }
            height="150"
            id="collection-img-id-mockCollectionId"
            patternUnits="userSpaceOnUse"
            width="150"
          />
        </defs>
        <rect
          fill="#e8e3eb"
          height="80"
          rx="5"
          width="80"
        />
        <path
          className="repository--collection-icon--pattern-path"
          d="m20.092289,37.657358q2.2001,-1.270228 4.400197,0l12.100543,6.986251q2.200097,1.270228 2.200097,3.810684l0,13.972502q0,2.540453 -2.200097,3.810681l-12.100543,6.986251q-2.200097,1.270228 -4.400197,0l-12.100543,-6.986251q-2.200097,-1.270228 -2.200097,-3.810681l0,-13.972502q0,-2.540456 2.200097,-3.810684l12.100543,-6.986251z"
          fill="url(#collection-img-id-mockCollectionId)"
        />
        <path
          d="m55.507514,37.65736q2.200094,-1.270229 4.400197,0l12.100545,6.986249q2.200094,1.270229 2.200094,3.81068l0,13.972507q0,2.54045 -2.200094,3.81068l-12.100545,6.986249q-2.200103,1.270229 -4.400197,0l-12.100545,-6.986249q-2.200094,-1.270229 -2.200094,-3.81068l0,-13.972507q0,-2.54045 2.200094,-3.81068l12.100545,-6.986249z"
          fill="#9482a6"
        />
        <path
          d="m37.861174,6.776266q2.200094,-1.270229 4.400197,0l12.100545,6.986249q2.200094,1.270229 2.200094,3.81068l0,13.972507q0,2.54045 -2.200094,3.81068l-12.100545,6.986249q-2.200103,1.270229 -4.400197,0l-12.100545,-6.986249q-2.200094,-1.270229 -2.200094,-3.81068l0,-13.972507q0,-2.54045 2.200094,-3.81068l12.100545,-6.986249z"
          fill="#b09ec3"
        />
      </svg>
    </div>
    <div
      className="repository--collection-icon--title"
    >
      Mock Collection Title
    </div>
  </a>
</div>
`;
